File: src/java/org/xmlBlaster/util/key/QueryKeyData.java
/*------------------------------------------------------------------------------
Name:      QueryKeyData.java
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
------------------------------------------------------------------------------*/
package org.xmlBlaster.util.key;

import org.xmlBlaster.util.Global;
import org.xmlBlaster.util.XmlBlasterException;
import org.xmlBlaster.util.def.Constants;
import org.xmlBlaster.util.def.ErrorCode;
import org.xmlBlaster.util.qos.AccessFilterQos;

import java.util.ArrayList;

/**
 * This class encapsulates the Message key information of query invocations. 
 * <p />
 * <ul>
 * <li>SubscribeKey Client side access facade</i>
 * <li>UnSubscribeKey Client side access facade</i>
 * <li>GetKey Client side access facade</i>
 * <li>EraseKey Client side access facade</i>
 * </ul>
 * <p>
 * For the xml representation see MsgKeySaxFactory.
 * </p>
 * @see org.xmlBlaster.util.key.QueryKeySaxFactory
 */
public final class QueryKeyData extends KeyData implements java.io.Serializable, Cloneable
{
   // NOTE: We parse all key attributes, but only a few are used for queries
   // e.g. contentMime and contentMimeExtended are not useful (the decorater classes hide them)
   private static final long serialVersionUID = 1L;
   private final static String ME = "QueryKeyData";
   private transient I_QueryKeyFactory factory;

   /**
    * subscribe(), get() and cluster configuration keys may contain a filter rule
    */
   protected ArrayList filterList = null;   // To collect the <filter> when sax parsing
   protected transient AccessFilterQos[] filterArr = null; // To cache the filters in an array

   /**
    * Minimal constructor.
    */
   public QueryKeyData(Global glob) {
      this(glob, (I_QueryKeyFactory)null, (String)null);
   }

   /**
    * @param glob
    * @param query The query string
    *        For example a topic oid like "Hello" or "oid:Hello"
    *        or a query like "xpath://key", "domain:CLUSTER"
    */
   public QueryKeyData(Global glob, String query) {
      this(glob);
      if (query == null || query.length() == 0) {
         throw new IllegalArgumentException("QueryKeyData got query=null argument");
      }

      if (query.startsWith("xpath:")) {
         this.queryType = Constants.XPATH;
         if (query.length() <= "xpath:".length())
            throw new IllegalArgumentException("QueryKeyData got query='xpath:' with no query");
         this.queryString = query.substring("xpath:".length());
      }
      else if (query.startsWith("subscriptionId:")) {
         this.queryType = Constants.EXACT;
         if (query.length() <= "subscriptionId:".length())
            throw new IllegalArgumentException("QueryKeyData got query='subscriptionId:' with empy id");
         setOid(query.substring("subscriptionId:".length()));
      }
      else if (query.startsWith("oid:")) {
         this.queryType = Constants.EXACT;
         if (query.length() <= "oid:".length())
            throw new IllegalArgumentException("QueryKeyData got query='oid:' with empy id");
         if (query.indexOf("\"") != -1 || query.indexOf("\'") != -1)
            throw new IllegalArgumentException("Please pass a valid topic oid without apostrophe \" or '");
         setOid(query.substring("oid:".length()));
      }
      else if (query.startsWith("domain:")) {
         this.queryType = Constants.EXACT;
         if (query.length() <= "domain:".length())
            throw new IllegalArgumentException("QueryKeyData got query='domain:' with empy id");
         setDomain(query.substring("domain:".length()));
      }
      else {
         this.queryType = Constants.EXACT;
         if (query.indexOf("\"") != -1 || query.indexOf("\'") != -1)
            throw new IllegalArgumentException("Please pass a valid topic oid without apostrophe \" or '");
         setOid(query);
      }
   }

   /**
    * @param glob
    * @param query The query string (syntax is depending on queryType)
    * @param queryType Constants.EXACT | Constants.XPATH | Constants.DOMAIN
    */
   public QueryKeyData(Global glob, String query, String queryType) throws XmlBlasterException {
      this(glob);
      if (query == null) {
         throw new IllegalArgumentException("QueryKeyData got query=null argument");
      }
      this.queryType = checkQueryType(queryType);
      if (isExact()) {
         setOid(query);
      }
      else if (isXPath()) {
         this.queryString = query;
      }
      else if (isDomain()) {
         setDomain(query);
      }
      else {
         throw new XmlBlasterException(glob, ErrorCode.INTERNAL_UNKNOWN, ME, "Your queryType=" + queryType + " is invalid, implementation is missing");
      }
   }

   /**
    * Constructor to parse a message. 
    * @param factory If null, the default factory from Global is used.
    */
   public QueryKeyData(Global glob, I_QueryKeyFactory factory, String serialData) {
      super(glob, serialData);
      this.factory = (factory == null) ? this.glob.getQueryKeyFactory() : factory;
   }

   /**
    * Check the query type, Constants.XPATH, Constants.EXACT, Constants.DOMAIN are currently supported
    */
   public String checkQueryType(String queryType) throws XmlBlasterException {
      queryType = queryType.toUpperCase();
      if (!Constants.EXACT.equals(queryType) &&
          !Constants.XPATH.equals(queryType) &&
          !Constants.DOMAIN.equals(queryType))
         throw new XmlBlasterException(glob, ErrorCode.USER_ILLEGALARGUMENT, ME,
              "Your queryType=" + queryType + " is invalid, use one of '"
               + Constants.EXACT + "' , '" + Constants.XPATH + "', '" + Constants.DOMAIN + "'");
      return queryType;
   }

   /**
    * Should be avoided to call directly.
    */
   public void setQueryType(String queryType) throws XmlBlasterException {
      this.queryType = checkQueryType(queryType);
      /*
      checkQueryType(queryType);
      String oldType = this.queryType;
      this.queryType = queryType.toUpperCase();
      if (!this.queryType.equals(oldType)) {
         if (isExact() && getOid()==null && oldType.equals(Constants.XPATH)) {
            super.setOid(this.queryString);
            this.queryString = null;
         }
         else if (isExact() && getOid()==null && oldType.equals(Constants.DOMAIN)) {
            super.setOid(getDomain());
            super.setDomain(null);
         }
         else if (isXPath() && this.queryString==null && oldType.equals(Constants.EXACT)) {
            this.queryString = getOid();
            //super.setOid(null);
         }
         else if (isXPath() && this.queryString==null && oldType.equals(Constants.DOMAIN)) {
            this.queryString = getDomain();
            super.setDomain(null);
         }
         else if (isDomain() && getDomain()==null && oldType.equals(Constants.EXACT)) {
            super.setDomain(getOid());
            //super.setOid(null);
         }
         else if (isDomain() && getDomain()==null && oldType.equals(Constants.XPATH)) {
            super.setDomain(this.queryString);
            this.queryString = null;
         }
      }
      */
   }

   /**
    * Use for domain specific query
    */
   public void setOid(String oid) {
      this.queryType = Constants.EXACT;
      super.setOid(oid);
   }

   /**
    * Use for domain specific query
    */
   public void setDomain(String domain) {
      // We don't force the queryType to DOMAIN here as this lets
      // the cluster tests e.g.
      //   java -Djava.compiler= junit.textui.TestRunner -noloading org.xmlBlaster.test.cluster.DirtyReadTest
      // fail. For clustering the domain='' is just an attribute used to
      // find a master.
      // On the other hand there is query support for DOMAIN in RequestBroker (in none cluster environment)
      // --> We need to resolve this!
      //this.queryType = Constants.DOMAIN;
      super.setDomain(domain);
   }

   /**
    * Your XPath query string. 
    * @param query The query string, e.g. "//key"
    */
   public void setQueryString(String query) {
      this.queryType = Constants.XPATH;
      this.queryString = query;
   }

   /**
    * Same as setQueryString() but allows you to call it more than once
    * the strings are concatenated. 
    * @param query The query string, e.g. "//key"
    */
   public void appendQueryString(String query) {
      this.queryType = Constants.XPATH;
      if (this.queryString == null) {
         this.queryString = query;
      }
      else {
         this.queryString += query;
      }
   }

   public String getQueryString() {
      return this.queryString;
   }

   /**
    * Helper which returns the oid OR the xpath query for nice logging. 
    */
   public String getNiceString() {
      if (getOid() != null && getOid().length() > 0) {
         return getOid();
      }
      return getQueryString();
   }

   /**
    * Return the filters or array with size==0 if none is specified. 
    * <p />
    * For subscribe() and get() and cluster messages.
    * @return never null
    */
   public synchronized AccessFilterQos[] getAccessFilterArr() {
      if (filterArr != null)
         return filterArr;

      if (filterList == null)
         return null;

      filterArr = new AccessFilterQos[filterList.size()];
      filterList.toArray(filterArr);
      return filterArr;
   }

   /*
    * For cluster master lookup only
    */
   public synchronized void addFilter(AccessFilterQos qos) {
      if (filterList == null) filterList = new ArrayList();
      filterList.add(qos);
      filterArr = null;
   }

   private I_QueryKeyFactory getFactory() {
      if (this.factory == null) {
         this.factory = this.glob.getQueryKeyFactory();
      }
      return this.factory;
   }

   /**
    * Converts the data in XML ASCII string.
    * @return An XML ASCII string
    */
   public String toXml() {
      return getFactory().writeObject(this, null);
   }

   /**
    * Dump state of this object into a XML ASCII string.
    * <br>
    * @param extraOffset indenting of tags for nice output
    * @return internal state of the query as a XML ASCII string
    */
   public String toXml(String extraOffset) {
      return getFactory().writeObject(this, extraOffset);
   }

   /**
    * Returns a shallow clone, you can change savely all basic or immutable types
    * like boolean, String, int.
    * Currently AccessFilterQos is not deep cloned (so don't change it)
    */
   public Object clone() {
      return super.clone();
   }
}

File: src/c++/util/key/QueryKeyData.cpp
/*------------------------------------------------------------------------------
Name:      QueryKeyData.cpp
Project:   xmlBlaster.org
Copyright: xmlBlaster.org, see xmlBlaster-LICENSE file
------------------------------------------------------------------------------*/

#include <util/key/QueryKeyData.h>
#include <util/Constants.h>
#include <util/XmlBlasterException.h>
#include <util/Global.h>
#include <algorithm>
#include <cctype>

namespace org { namespace xmlBlaster { namespace util { namespace key {

using namespace std;
using namespace org::xmlBlaster::util;
using namespace org::xmlBlaster::util::qos;

QueryKeyData::QueryKeyData(Global& global) : KeyData(global), accessFilterVector_()
{
}

QueryKeyData::QueryKeyData(Global& global, const string& query, const string& queryType) : KeyData(global), accessFilterVector_()
{
   this->queryType_ = checkQueryType(queryType);
   if (isExact()) {
      setOid(query);
   }
   else if (isXPath()) {
      this->queryString_ = query;
   }
   else if (isDomain()) {
      setDomain(query);
   }
   else {
      throw XmlBlasterException(USER_ILLEGALARGUMENT, ME + "::setQueryType",
                                "Your queryType=" + queryType_ + " is invalid, use one of '" + 
                                Constants::EXACT + "' , '" + Constants::XPATH + "' , '" + Constants::D_O_M_A_I_N + "'");
   }
}

QueryKeyData::QueryKeyData(const QueryKeyData& key) : KeyData(key), accessFilterVector_(key.accessFilterVector_)
{
}

QueryKeyData& QueryKeyData::operator =(const QueryKeyData& key) 
{
   accessFilterVector_ = key.accessFilterVector_;
   return *this;
}

string QueryKeyData::checkQueryType(const string& queryType)
{
   string tmp = queryType;
   // transform (tmp.begin(), tmp.end(), tmp.begin(), ::toupper);
   string::iterator iter = tmp.begin();
   while (iter != tmp.end()) {
      *iter = (char)::toupper(*iter);
      iter++;
   }

   if (Constants::EXACT != tmp && Constants::XPATH !=tmp && Constants::D_O_M_A_I_N !=tmp)
      throw XmlBlasterException(USER_ILLEGALARGUMENT, ME + "::setQueryType",
                                "Your queryType=" + queryType + " is invalid, use one of '" + 
                                Constants::EXACT + "' , '" + Constants::XPATH + "' , '" + Constants::D_O_M_A_I_N + "'");
   return tmp;
}

void QueryKeyData::setQueryType(const string& queryType)
{
   queryType_ = checkQueryType(queryType);
}

/**
 * Your XPath query string. 
 * NOTE: You need to set the query type first!
 * @param str Your tags in ASCII XML syntax
 */
void QueryKeyData::setQueryString(const string& tags)
{
   this->queryType_ = Constants::XPATH;
   this->queryString_ = tags;
}

void QueryKeyData::setOid(const string& oid)
{
   this->queryType_ = Constants::EXACT;
   oid_ = oid;
}

string QueryKeyData::getQueryString() const
{
   return queryString_;
}

/**
 * Return the filters or array with size==0 if none is specified. 
 * <p />
 * For subscribe() and get() and cluster messages.
 * @return never null
 */
AccessFilterVector QueryKeyData::getAccessFilterVector() const 
{
   return accessFilterVector_;
}

void QueryKeyData::addFilter(const AccessFilterQos& qos) 
{
   accessFilterVector_.insert(accessFilterVector_.end(), qos);
}

string QueryKeyData::toXml() const
{
   return toXml("");
}

string QueryKeyData::toXml(const string& extraOffset) const
{
   string ret;
   string offset = Constants::OFFSET + extraOffset;

   ret += offset + "<key oid='" + oid_ + "'";
   if (!getContentMime().empty())
      ret += " contentMime='" + getContentMime() + "'";
   if (!getContentMimeExtended().empty())
      ret += " contentMimeExtended='" + getContentMimeExtended() + "'";
   if (!getDomain().empty())
      ret += " domain='" + getDomain() + "'";

   if (!getQueryType().empty() && Constants::EXACT != getQueryType())
         ret += " queryType='" + getQueryType() + "'";
   ret += ">";
   if (!queryString_.empty()) {
      ret += offset + Constants::INDENT + getQueryString();
   }

   AccessFilterVector::const_iterator iter = accessFilterVector_.begin();
   while (iter != accessFilterVector_.end()) {
      ret += (*iter).toXml(extraOffset + Constants::INDENT);
      iter++;
   }
   ret += "</key>";
  return ret;
}

QueryKeyData* QueryKeyData::getClone() const
{
   return new QueryKeyData(*this);
}

}}}} // namespace

